From 0e3e21f756f23def7f49e0686f0ecfd284a1bb7d Mon Sep 17 00:00:00 2001
From: Justin <earthbound.iap@gmail.com>
Date: Sun, 5 Feb 2012 22:19:00 -0500
Subject: [PATCH 35/35] cpufreq: Prevent memory leak in cpufreq_stats on
 hotplug


diff --git a/Kernel/drivers/cpufreq/cpufreq_stats.c b/Kernel/drivers/cpufreq/cpufreq_stats.c
index 00d73fc..0b2b00c 100644
--- a/Kernel/drivers/cpufreq/cpufreq_stats.c
+++ b/Kernel/drivers/cpufreq/cpufreq_stats.c
@@ -305,6 +305,27 @@ static int cpufreq_stat_notifier_trans(struct notifier_block *nb,
 	return 0;
 }
 
+static int cpufreq_stats_create_table_cpu(unsigned int cpu)
+{
+  struct cpufreq_policy *policy;
+  struct cpufreq_frequency_table *table;
+  int ret = -ENODEV;
+
+  policy = cpufreq_cpu_get(cpu);
+  if (!policy)
+    return -ENODEV;
+
+  table = cpufreq_frequency_get_table(cpu);
+  if (!table)
+    goto out;
+
+  ret = cpufreq_stats_create_table(policy, table);
+
+out:
+  cpufreq_cpu_put(policy);
+  return ret;
+}
+
 static int __cpuinit cpufreq_stat_cpu_callback(struct notifier_block *nfb,
 					       unsigned long action,
 					       void *hcpu)
@@ -320,6 +341,10 @@ static int __cpuinit cpufreq_stat_cpu_callback(struct notifier_block *nfb,
 	case CPU_DEAD_FROZEN:
 		cpufreq_stats_free_table(cpu);
 		break;
+	case CPU_DOWN_FAILED:
+  	case CPU_DOWN_FAILED_FROZEN:
+    		cpufreq_stats_create_table_cpu(cpu);
+    		break;
 	}
 	return NOTIFY_OK;
 }
-- 
1.7.5.4

